// sp-rest-proxy/dist/gateway/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var SocketIOClient = require("socket.io-client");
var httpRequest = require("request");
var logger_1 = require("../utils/logger");
var Client = (function () {
    function Client(settings, proxy) {
        var _this = this;
        this.settings = settings;
        this.proxy = proxy;
        this.init = function () {
            _this.socket.on('REQUEST', function (request) {
                var endpoint = _this.enpointUrl(request.url);
                _this.logger.info(request.method + " request to " + endpoint);
                httpRequest(endpoint, {
                    method: request.method,
                    headers: request.headers
                }, function (err, response) {
                    var responsePackage = {
                        transaction: request.transaction,
                        err: err,
                        response: response
                    };
                    _this.socket.emit('RESPONSE', responsePackage);
                });
            });
        };
        this.enpointUrl = function (relativeUrl) {
            var hostname = _this.proxy.hostname;
            var port = _this.proxy.port;
            return ("http://" + hostname + ":" + port + relativeUrl)
                .replace(':80/', '/').replace(':443/', '/');
        };
        this.socket = SocketIOClient(settings.serverUrl);
        this.logger = new logger_1.Logger(proxy.logLevel);
    }
    return Client;
}());
exports.Client = Client;
